agent/hcp/telemetry_provider.go

```diff
@@ -0,0 +1,184 @@
+package hcp
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"regexp"
+	"sync"
+	"time"
+
+	goMetrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/mitchellh/hashstructure/v2"
+
+	hcpclient "github.com/hashicorp/consul/agent/hcp/client"
+	hcpTelemetry "github.com/hashicorp/consul/agent/hcp/telemetry"
+)
+
+var (
+	// internalMetricRefreshFailure is a metric to monitor refresh failures.
+	internalMetricRefreshFailure []string = []string{"hcp", "telemetry_config_provider", "refresh", "failure"}
+)
+
+// Ensure hcpProviderImpl implements telemetry provider interfaces.
+var _ hcpTelemetry.ConfigProvider = &hcpProviderImpl{}
+var _ hcpTelemetry.EndpointProvider = &hcpProviderImpl{}
```

Comment:
this feels a bit coupled to the telemetry configuration in `agent/hcp/client`, do you think this could be moved into there?

Reply:
This was added to force the HCP specific implementation (`hcpProviderImpl`) to conform to those interfaces. Partly documentation and partly ensuring fast at compile time the use case works

Those interfaces will be used by end users to provide their own implementation, so it needs to stay in the telemetry package.

Reply:
I understand the confirming to interface check bit, like it, do it a lot

I mean that this implementation is AFAICT the only implementation of those `ConfigProvider` and `EndpointProvider` interfaces. So it seems like this could be in `client` along-side the interfaces?

Reply:
Not sure I fully understand, happy to discuss sync to if that's better! 

Sorry do you mean moving this implementation, the actual implementation `hcpProviderImpl` and files `telemetry_provider*` into `hcp/client` package?

The interfaces, `ConfigProvider` and `EndpointProvider ` are in the `hcp/telemetry` package, and are required for the init of the OTELSink and OTELExporter. The interfaces would have to live there.

Reply:
> happy to discuss sync to if that's better!

sounds good to me, would be good to sync tomorrow

> Sorry do you mean moving this implementation, the actual implementation hcpProviderImpl and files telemetry_provider* into hcp/client package?

As you pointed out, I mean `hcp/telemetry`. So yeah, moving this implementation to alongside its interfaces

Reply:
We talked and part of the motivation is to keep `hcp/telemetry` free of HCP code so the port to `go-metrics` is easier down the line. If having these implementations outside that package help w/ getting it merged I'm in favor